pkg/cache/clusterqueue.go

```diff
@@ -493,3 +516,50 @@ func resetUsage(lqUsage FlavorResourceQuantities, cqUsage FlavorResourceQuantiti
 func workloadBelongsToLocalQueue(wl *kueue.Workload, q *kueue.LocalQueue) bool {
 	return wl.Namespace == q.Namespace && wl.Spec.QueueName == q.Name
 }
+
+// RequestableCohortQuota recalculates cohort.RequestableResources when LendingLimit enabled,
+// because when borrowing from the cohort, need to count cq's own nominal quota in.
+func (c *ClusterQueue) RequestableCohortQuota(fName kueue.ResourceFlavorReference, rName corev1.ResourceName) (val int64) {
+	if c.Cohort.RequestableResources == nil {
+		return 0
+	}
+	if c.Cohort.RequestableResources[fName] == nil {
+		return 0
+	}
+	requestableCohortQuota := c.Cohort.RequestableResources[fName][rName]
+
+	// When feature LendingLimit enabled, RequestableResources only contains the sum of
+	// all CQs' LendingLimit, so we need to add the guaranteed resources.
+	requestableCohortQuota += c.ReservedQuota(fName, rName)
+
+	return requestableCohortQuota
+}
+
+func (c *ClusterQueue) ReservedQuota(fName kueue.ResourceFlavorReference, rName corev1.ResourceName) (val int64) {
+	if !features.Enabled(features.LendingLimit) {
+		return 0
+	}
+	if c.GuaranteedQuota == nil {
+		return 0
+	}
+	if c.GuaranteedQuota[fName] == nil {
+		return 0
+	}
+	return c.GuaranteedQuota[fName][rName]
+}
+
+// CohortUsage recalculates cohort.Usage when LendingLimit enabled,
+// it represents the quota cq is using over its GuaranteedQuota.
+func (c *ClusterQueue) CohortUsage(fName kueue.ResourceFlavorReference, rName corev1.ResourceName) (val int64) {
+	cohortUsage := c.Cohort.Usage[fName][rName]
+
+	// When feature LendingLimit enabled, all CQs' GuaranteedQuota are reserved for itself, cannot be borrowed
+	// by other CQs in the cohort, so we need to remove the guaranteed resources from Cohort.Usage.
+	cohortUsage -= c.ReservedQuota(fName, rName)
```

Comment:
This is not right, because the cohortUsage you accumulated is the sum of all the clusterQueues, minus self guaranteed resource is not enough. You should process this for each clusterQueue.

Reply:
That's say you should process this in the snapshot for each clusterqueue, however, in CohortUsage, you should add the clusterQueue usage, the value is uncertained, it can be the usage if less then guaranteedQuota or the guaranteedQuota opposite.

Reply:
Done.